src/flint/types/nmod_poly.pyx

```diff
@@ -4,50 +4,103 @@ from flint.utils.typecheck cimport typecheck
 from flint.types.fmpz cimport fmpz, any_as_fmpz
 from flint.types.fmpz_poly cimport any_as_fmpz_poly
 from flint.types.fmpz_poly cimport fmpz_poly
-from flint.types.nmod cimport any_as_nmod
-from flint.types.nmod cimport nmod
+from flint.types.nmod cimport any_as_nmod_ctx
+from flint.types.nmod cimport nmod, nmod_ctx
 
 from flint.flintlib.nmod_vec cimport *
 from flint.flintlib.nmod_poly cimport *
 from flint.flintlib.nmod_poly_factor cimport *
 from flint.flintlib.fmpz_poly cimport fmpz_poly_get_nmod_poly
-from flint.flintlib.ulong_extras cimport n_gcdinv
+from flint.flintlib.ulong_extras cimport n_gcdinv, n_is_prime
 
 from flint.utils.flint_exceptions import DomainError
 
 
-cdef any_as_nmod_poly(obj, nmod_t mod):
-    cdef nmod_poly r
-    cdef mp_limb_t v
-    # XXX: should check that modulus is the same here, and not all over the place
-    if typecheck(obj, nmod_poly):
+_nmod_poly_ctx_cache = {}
+
+
+cdef nmod_poly_ctx any_as_nmod_poly_ctx(obj):
+    """Convert an int to an nmod_ctx."""
+    if typecheck(obj, nmod_poly_ctx):
         return obj
-    if any_as_nmod(&v, obj, mod):
-        r = nmod_poly.__new__(nmod_poly)
-        nmod_poly_init(r.val, mod.n)
-        nmod_poly_set_coeff_ui(r.val, 0, v)
-        return r
-    x = any_as_fmpz_poly(obj)
-    if x is not NotImplemented:
-        r = nmod_poly.__new__(nmod_poly)
-        nmod_poly_init(r.val, mod.n)   # XXX: create flint _nmod_poly_set_modulus for this?
-        fmpz_poly_get_nmod_poly(r.val, (<fmpz_poly>x).val)
-        return r
+    if typecheck(obj, int):
+        ctx = _nmod_poly_ctx_cache.get(obj)
+        if ctx is None:
+            ctx = nmod_poly_ctx(obj)
+            _nmod_poly_ctx_cache[obj] = ctx
+        return ctx
     return NotImplemented
 
-cdef nmod_poly_set_list(nmod_poly_t poly, list val):
-    cdef long i, n
-    cdef nmod_t mod
-    cdef mp_limb_t v
-    nmod_init(&mod, nmod_poly_modulus(poly)) # XXX
-    n = PyList_GET_SIZE(val)
-    nmod_poly_fit_length(poly, n)
-    for i from 0 <= i < n:
-        c = val[i]
-        if any_as_nmod(&v, val[i], mod):
-            nmod_poly_set_coeff_ui(poly, i, v)
-        else:
-            raise TypeError("unsupported coefficient in list")
+
+cdef nmod_poly nmod_poly_new_init(nmod_poly_ctx ctx):
+    cdef nmod_poly p
+    p = nmod_poly.__new__(nmod_poly)
+    nmod_poly_init(p.val, ctx.mod.n)
+    p.ctx = ctx
+    return p
+
+
+cdef nmod_poly nmod_poly_new_init_preinv(nmod_poly_ctx ctx):
+    cdef nmod_poly p
+    p = nmod_poly.__new__(nmod_poly)
+    nmod_poly_init_preinv(p.val, ctx.mod.n, ctx.mod.ninv)
+    p.ctx = ctx
+    return p
+
+
+cdef class nmod_poly_ctx:
+    """
+    Context object for creating :class:`~.nmod_poly` initalised 
+    with modulus :math:`N`.
+
+        >>> nmod_poly_ctx(17)
+        nmod_poly_ctx(17)
```

Comment:
We should have `__init__` raise an exception so you have to use something like `nmod_poly_ctx.new()`. Then we can make all contexts unique on construction.

Reply:
Done